### PR TITLE
Add sort order option, with new descending sort #1121

### PR DESCRIPTION
**Pull request by wrenoud**
Default behavior remains unchanged.
This includes unittests and documentation updates.

### CODE REVIEW

**Comment by wrenoud** [comment]
@abellgithub  I'll give it a try

File: filters/SortFilter.cpp [hunk 1]
```diff
@@ -51,49 +51,52 @@ void SortFilter::addArgs(ProgramArgs& args)
...
+        in.setstate(std::ios::failbit);
+    return in;
+}
+
+std::ostream& operator<<(std::ostream& out, const SortOrder& order)
+{
+    switch (order)
+    {
+    case SortOrder::ASC:
+        out << "Point";
```
Comment (abellgithub) [inline filters/SortFilter.cpp]: Text here is bad, but other than that, 👍 .  Thanks.

**Comment by wrenoud** [comment]
@abellgithub compiler asked for the default constructor, but I didn't entirely understand why:

```
2>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\algorithm(2565): error C2512: 'pdal::PointViewIter': no appropriate default constructor available (compiling source file PDAL\filters\SortFilter.cpp)
2>  PDAL\pdal/PointViewIter.hpp(99): note: see declaration of 'pdal::PointViewIter' (compiling source file PDAL\filters\SortFilter.cpp)
2>  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\algorithm(2879): note: see reference to function template instantiation 'void std::_Buffered_merge_unchecked<_BidIt,_Diff,pdal::PointIdxRef,_Pr>(_BidIt,_BidIt,_BidIt,_Diff,_Diff,std::_Temp_iterator<pdal::PointIdxRef> &,_Pr &)' being compiled
2>          with
2>          [
2>              _BidIt=pdal::PointViewIter,
2>              _Diff=__int64,
2>              _Pr=pdal::SortFilter::filter::<lambda_17181ad2d6bd7a35082218182bf9e741>
2>          ] (compiling source file PDAL\filters\SortFilter.cpp)
2>  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\algorithm(2891): note: see reference to function template instantiation 'void std::_Stable_sort_unchecked1<_BidIt,__int64,pdal::PointIdxRef,_Pr>(_BidIt,_BidIt,_Diff,std::_Temp_iterator<pdal::PointIdxRef> &,_Pr &)' being compiled
2>          with
2>          [
2>              _BidIt=pdal::PointViewIter,
2>              _Pr=pdal::SortFilter::filter::<lambda_17181ad2d6bd7a35082218182bf9e741>,
2>              _Diff=__int64
2>          ] (compiling source file PDAL\filters\SortFilter.cpp)
2>  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\algorithm(2900): note: see reference to function template instantiation 'void std::_Stable_sort_unchecked<_Iter,_Pr>(_BidIt,_BidIt,_Pr &)' being compiled
2>          with
2>          [
2>              _Iter=pdal::PointViewIter,
2>              _Pr=pdal::SortFilter::filter::<lambda_17181ad2d6bd7a35082218182bf9e741>,
2>              _BidIt=pdal::PointViewIter
2>          ] (compiling source file PDAL\filters\SortFilter.cpp)
2>  PDAL\filters\SortFilter.cpp(72): note: see reference to function template instantiation 'void std::stable_sort<pdal::PointViewIter,pdal::SortFilter::filter::<lambda_17181ad2d6bd7a35082218182bf9e741>>(_BidIt,_BidIt,_Pr)' being compiled
2>          with
2>          [
2>              _BidIt=pdal::PointViewIter,
2>              _Pr=pdal::SortFilter::filter::<lambda_17181ad2d6bd7a35082218182bf9e741>
2>          ]
```
The other bug where I added const to the difference operator visual studio didn't complain, it was just gcc that complained about const'ness being cast away: 
```
In file included from /usr/include/c++/5/algorithm:62:0,
                 from /pdal/pdal/util/Utils.hpp:37,
                 from /pdal/pdal/pdal_types.hpp:46,
                 from /pdal/pdal/pdal_internal.hpp:43,
                 from /pdal/pdal/Stage.hpp:39,
                 from /pdal/pdal/Filter.hpp:37,
                 from /pdal/filters/SortFilter.hpp:37,
                 from /pdal/filters/SortFilter.cpp:35:
/usr/include/c++/5/bits/stl_algo.h: In instantiation of 'void std::__stable_sort_adaptive(_RandomAccessIterator, _RandomAccessIterator, _Pointer, _Distance, _Compare) [with _RandomAccessIterator = pdal::PointViewIter; _Pointer = pdal::PointIdxRef*; _Distance = long int; _Compare = __gnu_cxx::__ops::_Iter_comp_iter<pdal::SortFilter::filter(pdal::PointView&)::<lambda(const pdal::PointIdxRef&, const pdal::PointIdxRef&)> >]':
/usr/include/c++/5/bits/stl_algo.h:4865:29:   required from 'void std::__stable_sort(_RandomAccessIterator, _RandomAccessIterator, _Compare) [with _RandomAccessIterator = pdal::PointViewIter; _Compare = __gnu_cxx::__ops::_Iter_comp_iter<pdal::SortFilter::filter(pdal::PointView&)::<lambda(const pdal::PointIdxRef&, const pdal::PointIdxRef&)> >]'
/usr/include/c++/5/bits/stl_algo.h:4932:36:   required from 'void std::stable_sort(_RAIter, _RAIter, _Compare) [with _RAIter = pdal::PointViewIter; _Compare = pdal::SortFilter::filter(pdal::PointView&)::<lambda(const pdal::PointIdxRef&, const pdal::PointIdxRef&)>]'
/pdal/filters/SortFilter.cpp:72:51:   required from here
/usr/include/c++/5/bits/stl_algo.h:2762:27: error: no match for 'operator-' (operand types are 'const pdal::PointViewIter' and 'pdal::PointViewIter')
        _Distance(__middle - __first),
                           ^
In file included from /pdal/filters/SortFilter.hpp:38:0,
                 from /pdal/filters/SortFilter.cpp:35:
/pdal/pdal/PointViewIter.hpp:137:21: note: candidate: pdal::PointViewIter::difference_type pdal::PointViewIter::operator-(const pdal::PointViewIter&) <near match>
     difference_type operator-(const PointViewIter& i)
                     ^
/pdal/pdal/PointViewIter.hpp:137:21: note:   passing 'const pdal::PointViewIter*' as 'this' argument discards qualifiers
```
It was that last line that gave me the hint.

**Comment by hobu** [comment]
ready for merge?

**Comment by wrenoud** [comment]
@hobu I was just waiting to see if @abellgithub had more feedback. The one outstanding question was around performance with std::stable_sort. I did a cursory search for existing performance benchmarks. 

[This](https://solarianprogrammer.com/2012/10/24/cpp-11-sort-benchmark/) is fairly old but the most comprehesive review of performance across compilers, showing Clang as the only compiler that is a clear winner for `std::sort`.

Another option could be added to allow the user to select stable sort.

**Comment by abellgithub** [comment]
On Mon, Mar 27, 2017 at 5:01 AM, Weston Renoud <notifications@github.com>
wrote:

> @hobu <https://github.com/hobu> I was just waiting to see if @abellgithub
> <https://github.com/abellgithub> had more feedback. The one outstanding
> question was around performance with std::stable_sort. I did a cursory
> search for existing performance benchmarks.
>
> This <https://solarianprogrammer.com/2012/10/24/cpp-11-sort-benchmark/>
> is fairly old but the most comprehesive review of performance across
> compilers, showing Clang as the only compiler that is a clear winner for
> std::sort.
>

Don't know that I think much of the test, but if we want to change to
stable sort and close the ticket for now, that's fine.  I don't think the
filter is heavily used and if there's an issue, we can address later.

-- 
Andrew Bell
andrew.bell.ia@gmail.com

**Pull request merged**